examples/tutorials/core_api_mnist/model_def_metrics.py

```diff
@@ -0,0 +1,169 @@
+from __future__ import print_function
+import argparse
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+import torch.optim as optim
+from torchvision import datasets, transforms
+from torch.optim.lr_scheduler import StepLR
+
+# NEW: imports 
+import determined as det
+import pathlib
+
+class Net(nn.Module):
+    def __init__(self):
+        super(Net, self).__init__()
+        self.conv1 = nn.Conv2d(1, 32, 3, 1)
+        self.conv2 = nn.Conv2d(32, 64, 3, 1)
+        self.dropout1 = nn.Dropout(0.25)
+        self.dropout2 = nn.Dropout(0.5)
+        self.fc1 = nn.Linear(9216, 128)
+        self.fc2 = nn.Linear(128, 10)
+
+    def forward(self, x):
+        x = self.conv1(x)
+        x = F.relu(x)
+        x = self.conv2(x)
+        x = F.relu(x)
+        x = F.max_pool2d(x, 2)
+        x = self.dropout1(x)
+        x = torch.flatten(x, 1)
+        x = self.fc1(x)
+        x = F.relu(x)
+        x = self.dropout2(x)
+        x = self.fc2(x)
+        output = F.log_softmax(x, dim=1)
+        return output
+
+def train(args, model, device, train_loader, optimizer, epoch, core_context):
+    model.train()
+    for batch_idx, (data, target) in enumerate(train_loader):
+        data, target = data.to(device), target.to(device)
+        optimizer.zero_grad()
+        output = model(data)
+        loss = F.nll_loss(output, target)
+        loss.backward()
+        optimizer.step()
+        if (batch_idx + 1) % args.log_interval == 0:
+            
+            print('Train Epoch: {} [{}/{} ({:.0f}%)]\tLoss: {:.6f}'.format(
+            epoch, (batch_idx) * len(data), len(train_loader.dataset),
+            100. * (batch_idx) / len(train_loader), loss.item()))
+            
+            # NEW: report training metrics to Determined master via core context. 
+            # Index by (batch_idx + 1) * (epoch-1) * len(train_loader) to continuously plot loss on one graph for consecutive epochs.
+            core_context.train.report_training_metrics(steps_completed=(batch_idx + 1) + (epoch-1) * len(train_loader), \
+            metrics={"train_loss":loss.item()})
+      
+            if args.dry_run:
+                break
+
+
+# NEW: modified function header
```

Comment:
I'd suggest explaining what changes you made and why

Reply:
Also, for this `test` function, I'd suggest just adding a `steps_completed` parameter, rather than injecting the information needed to calculate steps_completed.

My reasoning is that it's a bit easier to explain: "you need steps_completed to report validation metrics" vs "you need the length of the training dataset so you can combine it with the epoch so you can get the steps_completed so you can report validation metrics"